eng/common/scripts/get-codeowners.ps1

```diff
@@ -1,49 +1,89 @@
 param (
-  $TargetDirectory, # should be in relative form from root of repo. EG: sdk/servicebus
-  $RootDirectory, # ideally $(Build.SourcesDirectory)
-  $VsoVariable = "" # target devops output variable
+  [string]$TargetDirectory = "", # Code path to code owners. e.g sdk/core/azure-amqp
+  [string]$RootDirectory = "$env:SYSTEM_DEFAULTWORKINGDIRECTORY", # The repo contains CODEOWNER file.
+  [string]$CodeOwnerFileLocation = "$PSSCriptRoot/../../../.github/CODEOWNERS", # The absolute path of CODEOWNERS file. 
+  [string]$ToolVersion = "1.0.0-dev.20211118.20", # Placeholder. Will update in next PR
+  [string]$ToolPath = (Join-Path ([System.IO.Path]::GetTempPath()) "codeowners-tool-path"), # The place to check the tool existence. Put temp path as default
+  [string]$DevOpsFeed = "https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json", # DevOp tool feeds.
+  [string]$VsoVariable = "", # Option of write code owners into devop variable
+  [switch]$Test  #Run test functions against the script logic
 )
-$target = $TargetDirectory.ToLower().Trim("/")
-$codeOwnersLocation = Join-Path $RootDirectory -ChildPath ".github/CODEOWNERS"
-$ownedFolders = @{}
 
-if (!(Test-Path $codeOwnersLocation)) {
-  Write-Host "Unable to find CODEOWNERS file in target directory $RootDirectory"
-  exit 1
-}
-
-$codeOwnersContent = Get-Content $codeOwnersLocation
+function Get-CodeOwnersTool()
+{
+  $command = Join-Path $ToolPath "retrieve-codeowners"
+  # Check if the retrieve-codeowners tool exsits or not.
+  if (Get-Command $command -errorAction SilentlyContinue) {
+    return $command
+  }
+  if (!(Test-Path $ToolPath)) {
+    New-Item -ItemType Directory -Path $ToolPath | Out-Null
+  }
+  Write-Host "Installing the retrieve-codeowners tool under $ToolPath... "
+  dotnet tool install --tool-path $ToolPath --add-source $DevOpsFeed --version $ToolVersion "Azure.Sdk.Tools.RetrieveCodeOwners" | Out-Null
 
-foreach ($contentLine in $codeOwnersContent) {
-  if (-not $contentLine.StartsWith("#") -and $contentLine){
-    $splitLine = $contentLine -split "\s+"
-    
-    # CODEOWNERS file can also have labels present after the owner aliases
-    # gh aliases start with @ in codeowners. don't pass on to API calls
-    $ownedFolders[$splitLine[0].ToLower().Trim("/")] = ($splitLine[1..$($splitLine.Length)] `
-      | ? { $_.StartsWith("@") } `
-      | % { return $_.substring(1) }) -join ","
+  # Test to see if the tool properly installed.
+  if (!(Get-Command $command -errorAction SilentlyContinue)) {
+    Write-Error "The retrieve-codeowners tool is not properly installed. Please check your tool path. $ToolPath"
+    return 
   }
+  return $command
 }
 
-$results = $ownedFolders[$target]
-
-if ($results) {
-  Write-Host "Found a folder $results to match $target"
+function Get-CodeOwners ([string]$targetDirectory = $TargetDirectory, [string]$codeOwnerFileLocation = $CodeOwnerFileLocation)
```

Comment:
```suggestion
function Get-CodeOwners ([string]$targetDirectory, [string]$codeOwnerFileLocation)
```
Let's not default those parameters but instead just pass them in the place you are calling this function below. The reason for that is we want to avoid conflicts.